src/main/host/process.rs

```diff
@@ -1551,4 +1557,143 @@ mod export {
+                None => std::ptr::null_mut(),
+            }
+        })
+        .unwrap()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn _process_removeThread(proc: *const RustProcess, tid: libc::pid_t) {
+        let proc = unsafe { proc.as_ref().unwrap() };
+        Worker::with_active_host(|host| {
+            let proc = proc.borrow(host.root());
+            let tid = ThreadId::try_from(tid).unwrap();
+            proc.threads.borrow_mut().remove(&tid);
+        })
+        .unwrap()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn _process_insertThread(
+        proc: *const RustProcess,
+        thread: *mut cshadow::Thread,
+    ) {
+        let proc = unsafe { proc.as_ref().unwrap() };
+        Worker::with_active_host(|host| {
+            let proc = proc.borrow(host.root());
+            let thread = unsafe { ThreadRef::new(thread) };
+            let tid = thread.id();
+            proc.threads.borrow_mut().insert(tid, thread);
+        })
+        .unwrap()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn _process_isRunning(proc: *const RustProcess) -> bool {
+        let proc = unsafe { proc.as_ref().unwrap() };
+        Worker::with_active_host(|host| {
+            let proc = proc.borrow(host.root());
+            !proc.is_exiting.get() && proc.threads.borrow().len() > 0
+        })
+        .unwrap()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn _process_numThreads(proc: *const RustProcess) -> usize {
+        let proc = unsafe { proc.as_ref().unwrap() };
+        Worker::with_active_host(|host| {
+            let proc = proc.borrow(host.root());
+            let threads = proc.threads.borrow();
+            threads.len()
+        })
+        .unwrap()
+    }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn _process_interruptWithSignal(
+        proc: *const RustProcess,
+        host_lock: *mut ShimShmemHostLock,
+        signo: i32,
+    ) {
+        let proc = unsafe { proc.as_ref().unwrap() };
+        let host_lock = unsafe { host_lock.as_mut().unwrap() };
+        Worker::with_active_host(|host| {
+            let proc = proc.borrow(host.root());
+            let signal = Signal::try_from(signo).unwrap();
+            let host_shmem = host.shim_shmem_lock_borrow().unwrap();
+            let mut threads = proc.threads.borrow_mut();
+            for (_tid, thread) in threads.iter_mut() {
+                {
+                    let thread_shmem = thread.shmem();
+                    let thread_shmem_protected =
+                        thread_shmem.protected.borrow_mut(&host_shmem.root);
+                    let blocked_signals = thread_shmem_protected.blocked_signals;
+                    if blocked_signals.has(signal) {
+                        continue;
+                    }
+                }
+                let Some(mut cond) = thread.syscall_condition_mut() else {
+                    continue;
+                };
+                cond.wakeup_for_signal(host_lock, signal);
```

Comment:
Maybe add a comment for what the condition is where there is no syscall condition? Shouldn't all running threads be blocked on some syscall?

Reply:
Yeah, I *think* it shouldn't happen. I tried adding a `panic` here, and none of the tests triggered it. I'm not  quite confident enough to leave that in though, so added a comment and logging a warning instead.